streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -1543,26 +1544,28 @@ private boolean close(final Optional<Long> timeout, final boolean leaveGroup) {
             timeoutMs = Long.MAX_VALUE;
         }
 
-        if (state.hasCompletedShutdown()) {
-            log.info("Streams client is already in the terminal {} state, all resources are closed and the client has stopped.", state);
-            return true;
-        }
-        if (state.isShuttingDown()) {
-            log.info("Streams client is in {}, all resources are being closed and the client will be stopped.", state);
-            if (state == State.PENDING_ERROR && waitOnState(State.ERROR, timeoutMs)) {
-                log.info("Streams client stopped to ERROR completely");
-                return true;
-            } else if (state == State.PENDING_SHUTDOWN && waitOnState(State.NOT_RUNNING, timeoutMs)) {
-                log.info("Streams client stopped to NOT_RUNNING completely");
+        if (!setState(State.PENDING_SHUTDOWN)) {
+
+            if (state.isShuttingDown()) {
```

Comment:
We know that are in a shutting-down or shut-down state if the transition to `PENDING_SHUTDOWN` fails, so either `isShuttingDown` or `hasCompletedShutdown` will be true. However, it is important that we check `isShuttingDown` before `hasCompletedShutdown`, as we otherwise can get a race condition where both return false (if the other thread transitions from `PENDING_SHUTDOWN` to `SHUTDOWN` in the meantime).

Reply:
Not sure if we still need/want this? Why do we care if we are in "pending" state, or already fully terminated? In the end, we could not transit to "pending_shotdown", and just need to wait for the terminal state. What do we gain to keep this code path (also cf below: I think it's error prone anyway).